fullstory-segment-middleware/src/test/java/com/fullstorydev/fullstory_segment_middleware/FullStorySegmentMiddlewareTest.java

```diff
@@ -0,0 +1,442 @@
+
+        PowerMockito.mockStatic(FS.class);
+        when(FS.getCurrentSessionURL(anyBoolean())).thenReturn("mock FullStory Session URL");
+        when(FS.getCurrentSessionURL()).thenReturn("mock FullStory Session URL");
+    }
+
+    @Test
+    public void constructor_WithAllowListedEvent() {
+        ArrayList<String> allowList = new ArrayList<>();
+        allowList.add("Product Added");
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey", allowList);
+        Assert.assertEquals(allowList, fullStorySegmentMiddleware.allowlistedEvents);
+        // TODO: assert logging for SharedPreferences listener code
+    }
+
+    @Test
+    public void constructor_NoAllowListedEvent() {
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey");
+        Assert.assertTrue(fullStorySegmentMiddleware.enableFSSessionURLInEvents);
+        Assert.assertFalse(fullStorySegmentMiddleware.enableGroupTraitsAsUserVars);
+        Assert.assertFalse(fullStorySegmentMiddleware.enableSendScreenAsEvents);
+        Assert.assertFalse(fullStorySegmentMiddleware.allowlistAllTrackEvents);
+        Assert.assertTrue(fullStorySegmentMiddleware.allowlistedEvents.isEmpty());
+        // TODO: assert logging for null allowlist warning
+    }
+
+    @Test
+    public void intercept_GroupPayloadChain_DisableGroupTraitsAsUserVars_ChainProceedCalled() {
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey");
+        fullStorySegmentMiddleware.enableGroupTraitsAsUserVars = false;
+        fullStorySegmentMiddleware.enableFSSessionURLInEvents = false;
+        GroupPayload groupPayload = new GroupPayload.Builder().userId("userId").groupId("groupId").build();
+        when(mockChain.payload()).thenReturn(groupPayload);
+        fullStorySegmentMiddleware.intercept(mockChain);
+
+        // chain.proceed is called with the correct payload
+        verify(mockChain, times(1)).proceed(groupPayload);
+    }
+
+    @Test
+    public void intercept_GroupPayloadChain_DisableGroupTraitsAsUserVars_FSSetUserVarsCalled() {
+        fullStorySegmentMiddleware = new FullStorySegmentMiddleware(mockContext, "SegmentWriteMockKey");
+        fullStorySegmentMiddleware.enableGroupTraitsAsUserVars = false;
+        fullStorySegmentMiddleware.enableFSSessionURLInEvents = false;
+        GroupPayload groupPayload = new GroupPayload.Builder().userId("userId").groupId("groupId").build();
+        when(mockChain.payload()).thenReturn(groupPayload);
+        fullStorySegmentMiddleware.intercept(mockChain);
+
+        // FS.event is called with properties
+        Map<String, String> userVars = new HashMap<>();
+        userVars.put("groupID_str", groupPayload.groupId());
+        verifyStatic(FS.class, VerificationModeFactory.times(1));
+        FS.setUserVars(userVars);
```

Comment:
Do you need to invoke `setUserVars` here?

Reply:
`verifyStatic` is a less than intuitive method, line 110 and 111 basically says verify `FS.setUserVars` is invoked 1 time with `userVars` as input: 
https://www.javadoc.io/static/org.powermock/powermock-api-mockito/1.7.0/org/powermock/api/mockito/PowerMockito.html#verifyStatic-org.mockito.verification.VerificationMode-

Reply:
I was thinking from the title of the test that the test logic was something like: "when group traits are disabled on the  middleware, the middleware interceptor calls `FS.setUserVars`" 

but it seems like the test logic is more like "when I call `FS.setUserVars`, verify that it is invoked once"

Am I reading this correctly?

Reply:
Ahh thanks I see what you're saying now... So in order for `verifyStatic` to work, it needs to know which static methods I'm verifying - by actually calling it immediately after(which is not counted towards the times of invocation). 
So it would work something like this:
```
// pass:
FS.setUserVars(userVars);
FS.setUserVars(userVars);
FS.setUserVars(userVars);
verifyStatic(FS.class, VerificationModeFactory.times(3));
FS.setUserVars(userVars);

// won't pass:
verifyStatic(FS.class, VerificationModeFactory.times(3));
FS.setUserVars(userVars);
FS.setUserVars(userVars);
FS.setUserVars(userVars);

// won't pass:
FS.setUserVars(userVars);
FS.setUserVars(userVars);
FS.setUserVars(userVars);
verifyStatic(FS.class, VerificationModeFactory.times(3));
```

I think this might a better example [here](https://github.com/powermock/powermock/wiki/mockito#a-full-example-for-mocking-stubbing--verifying-static-method)

Reply:
I talked with Sabrina and this statement is working as expected.